sql/src/main/java/org/apache/druid/sql/http/SqlResource.java

```diff
@@ -34,72 +34,121 @@
   public static final String SQL_HEADER_VALUE = "yes";
+
   private static final Logger log = new Logger(SqlResource.class);
-  public static final SqlResourceQueryMetricCounter QUERY_METRIC_COUNTER = new SqlResourceQueryMetricCounter();
+  private static final SqlResourceQueryMetricCounter QUERY_METRIC_COUNTER = new SqlResourceQueryMetricCounter();
 
   private final ObjectMapper jsonMapper;
   private final AuthorizerMapper authorizerMapper;
-  private final SqlStatementFactory sqlStatementFactory;
-  private final SqlLifecycleManager sqlLifecycleManager;
   private final ServerConfig serverConfig;
   private final ResponseContextConfig responseContextConfig;
   private final DruidNode selfNode;
+  private final SqlLifecycleManager sqlLifecycleManager;
+  private final SqlEngineRegistry sqlEngineRegistry;
 
+  @VisibleForTesting
   @Inject
-  protected SqlResource(
+  public SqlResource(
       final ObjectMapper jsonMapper,
       final AuthorizerMapper authorizerMapper,
-      final @NativeQuery SqlStatementFactory sqlStatementFactory,
-      final SqlLifecycleManager sqlLifecycleManager,
       final ServerConfig serverConfig,
+      final SqlLifecycleManager sqlLifecycleManager,
+      final SqlEngineRegistry sqlEngineRegistry,
       ResponseContextConfig responseContextConfig,
       @Self DruidNode selfNode
   )
   {
+    this.sqlEngineRegistry = Preconditions.checkNotNull(sqlEngineRegistry, "sqlEngineRegistry");
     this.jsonMapper = Preconditions.checkNotNull(jsonMapper, "jsonMapper");
     this.authorizerMapper = Preconditions.checkNotNull(authorizerMapper, "authorizerMapper");
-    this.sqlStatementFactory = Preconditions.checkNotNull(sqlStatementFactory, "sqlStatementFactory");
-    this.sqlLifecycleManager = Preconditions.checkNotNull(sqlLifecycleManager, "sqlLifecycleManager");
     this.serverConfig = Preconditions.checkNotNull(serverConfig, "serverConfig");
     this.responseContextConfig = responseContextConfig;
     this.selfNode = selfNode;
+    this.sqlLifecycleManager = Preconditions.checkNotNull(sqlLifecycleManager, "sqlLifecycleManager");
+  }
+
+  @GET
+  @Path("/engines")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSupportedEngines(@Context final HttpServletRequest request)
+  {
+    AuthorizationUtils.setRequestAuthorizationAttributeIfNeeded(request);
+    return Response.ok(new SupportedEnginesResponse(sqlEngineRegistry.getSupportedEngines())).build();
+  }
+
+  /**
+   * API to list all running queries, for an engine that supports such listings.
+   *
+   * @param selfOnly if true, return queries running on this server. If false, return queries running on all servers.
+   * @param engine engine name.
+   * @param request  http request.
+   */
+  @GET
+  @Path("/queries")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response doGetRunningQueries(
+      @QueryParam("selfOnly") final String selfOnly,
+      @QueryParam("engine") @Nullable final String engine,
```

Comment:
IMO, this API should not have an `engine` parameter. It should be specced to return all running queries for all engines that support the "get running queries" operation. It's ok if it does not return any native queries for now.